### PR TITLE
linux: update to 5.16.y

### PR DESCRIPTION
**Pull request by heitbaum**
https://cdn.kernel.org/pub/linux/kernel/v5.x/ChangeLog-5.16

### errors / fixes / issues / regressions
**Fixes**
- Bluetooth: Read codec capabilities only if supported (patch included) - https://github.com/LibreELEC/LibreELEC.tv/pull/6103/commits/714ea98fc61ea3d49a2fe20dfd4d6eb5330b7140
- Exynos : mali-midgard updated to https://github.com/LibreELEC/mali-midgard/commit/2c2accf67356463ee661627b8705429256011dcb
- nvidia-legacy: https://github.com/LibreELEC/LibreELEC.tv/pull/6103/commits/5c8e03fe3195f51d2738cd264ee1223dd6f01fe0
- brcmfmac_sdio-firmware: add and update symlinks for Cubox-i 4405ca8a1adfcdaa41027a20cee658cc059adfbf and a12b947a3976981cf276e99b531e3f0a702ece7d

### log 
- https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/log/?h=linux-5.16.y
- https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable-rc.git/log/?h=linux-5.16.y
- https://git.kernel.org/pub/scm/linux/kernel/git/stable/stable-queue.git/log/
- https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable-rc.git/log/?h=queue/5.16

### 5.16 Build tested on all of:

```
PROJECT=Allwinner ARCH=arm DEVICE=A64 s/build linux
PROJECT=Allwinner ARCH=arm DEVICE=H3 s/build linux
PROJECT=Allwinner ARCH=arm DEVICE=H5 s/build linux
PROJECT=Allwinner ARCH=arm DEVICE=H6 s/build linux
PROJECT=Rockchip ARCH=arm DEVICE=RK3288 s/build linux
PROJECT=Rockchip ARCH=arm DEVICE=RK3328 s/build linux
PROJECT=Rockchip ARCH=arm DEVICE=RK3399 s/build linux
PROJECT=NXP ARCH=arm DEVICE=iMX6 s/build linux
PROJECT=NXP ARCH=arm DEVICE=iMX8 s/build linux
PROJECT=Qualcomm ARCH=arm DEVICE=Dragonboard s/build linux
PROJECT=Samsung ARCH=arm DEVICE=Exynos s/build linux
PROJECT=Generic ARCH=x86_64 DEVICE=Generic s/build linux
PROJECT=Generic ARCH=x86_64 DEVICE=Generic-legacy s/build linux
```

### 5.16.y Run tested on all of:
- Allwinner all - tested - TBA - jernejsk
- Allwinner H6 (Tanix TX6) - TBA  - heitbaum
- Generic Generic (Intel TGL - NUC11) - 5.16 - heitbaum
- NXP iMX6 (Cubox-i4Pro) - 5.16  - heitbaum
- Rockchip RK3399pro (Rock Pi N10) - TBA - heitbaum
- RK all - tested - TBA - knaerzche
- Samsung Exynos (Hardkernel ODROID XU4) - TBA - heitbaum (video issue)

### CODE REVIEW

File: projects/Allwinner/patches/linux/0040-a64-increase-mali-frequency.patch [hunk 1]
```diff
@@ -11,9 +11,9 @@ Signed-off-by: Jernej Skrabec <jernej.skrabec@siol.net>
 --- a/arch/arm64/boot/dts/allwinner/sun50i-a64.dtsi
 +++ b/arch/arm64/boot/dts/allwinner/sun50i-a64.dtsi
 @@ -1141,6 +1141,9 @@
- 			clocks = <&ccu CLK_BUS_GPU>, <&ccu CLK_GPU>;
  			clock-names = "bus", "core";
  			resets = <&ccu RST_BUS_GPU>;
+ 			operating-points-v2 = <&gpu_opp_table>;
 +
 +			assigned-clocks = <&ccu CLK_GPU>;
 +			assigned-clock-rates = <432000000>;
```
Comment (jernejsk) [inline projects/Allwinner/patches/linux/0040-a64-increase-mali-frequency.patch:19]: This patch should be removed. Operating points replaces assigned clock rate property.

Comment (heitbaum) [reply projects/Allwinner/patches/linux/0040-a64-increase-mali-frequency.patch]: 👍 done

**Comment by heitbaum** [comment]
All done (except for the testing)

**Comment by heitbaum** [comment]
### SolidRun Cubox-i Dual/Quad doesn’t find the firmware at first.
```
cubox:~ # dmesg | head -2; dmesg | grep -i -e cubox -e brcmf
[    0.000000] Booting Linux on physical CPU 0x0
[    0.000000] Linux version 5.16.0 (docker@c7e15a8b922d) (armv7a-libreelec-linux-gnueabihf-gcc-10.3.0 (GCC) 10.3.0, GNU ld (GNU Binutils) 2.35.1) #1 SMP Mon Jan 10 08:28:47 UTC 2022
[    0.000000] OF: fdt: Machine model: SolidRun Cubox-i Dual/Quad
[   16.516056] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac4330-sdio for chip BCM4330/4
[   16.519471] brcmfmac mmc0:0001:1: Direct firmware load for brcm/brcmfmac4330-sdio.solidrun,cubox-i-q.bin failed with error -2
[   16.968659] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac4330-sdio for chip BCM4330/4
[   16.969275] brcmfmac: brcmf_c_process_clm_blob: no clm_blob available (err=-2), device may have limited channels available
[   16.970633] brcmfmac: brcmf_c_preinit_dcmds: Firmware: BCM4330/4 wl0: Oct 25 2011 19:34:12 version 5.90.125.104
```

**Comment by chewitt** [comment]
> brcmfmac4330-sdio.solidrun,cubox-i-q.bin

Harmless, but can be silenced with an alias in the firmware repo.

**Comment by heitbaum** [comment]
> > brcmfmac4330-sdio.solidrun,cubox-i-q.bin
> 
> Harmless, but can be silenced with an alias in the firmware repo.

After the rebuild referencing the new package - the .bin links did not build.

```
cubox:~ # find / -name brcmfmac4329-sdio.solidrun,cubox-i-q.txt
/run/kernel-overlays/firmware/brcm/brcmfmac4329-sdio.solidrun,cubox-i-q.txt
/usr/lib/kernel-overlays/base/lib/firmware/brcm/brcmfmac4329-sdio.solidrun,cubox-i-q.txt
cubox:~ # find / -name brcmfmac4329-sdio.solidrun,cubox-i-q.bin
cubox:~ # find / -name brcmfmac4330-sdio.solidrun,cubox-i-q.txt
/run/kernel-overlays/firmware/brcm/brcmfmac4330-sdio.solidrun,cubox-i-q.txt
/usr/lib/kernel-overlays/base/lib/firmware/brcm/brcmfmac4330-sdio.solidrun,cubox-i-q.txt
cubox:~ # find / -name brcmfmac4330-sdio.solidrun,cubox-i-q.bin
cubox:~ # find / | grep brcmfmac4330-sdio | grep .bin$
/run/kernel-overlays/firmware/brcm/brcmfmac4330-sdio.bin
/usr/lib/kernel-overlays/base/lib/firmware/brcm/brcmfmac4330-sdio.bin
cubox:~ # ls -la /usr/lib/kernel-overlays/base/lib/firmware/brcm/brcmfmac4330*
-rw-r--r--    1 root     root          1509 Jan 10 05:43 /usr/lib/kernel-overlays/base/lib/firmware/brcm/brcmfmac4330-sdio.Prowise-PT301.txt
-rw-r--r--    1 root     root        222126 Jan 10 05:43 /usr/lib/kernel-overlays/base/lib/firmware/brcm/brcmfmac4330-sdio.bin
-rw-r--r--    1 root     root          1789 Jan 10 21:15 /usr/lib/kernel-overlays/base/lib/firmware/brcm/brcmfmac4330-sdio.solidrun,cubox-i-dl.txt
-rw-r--r--    1 root     root          1789 Jan 10 21:15 /usr/lib/kernel-overlays/base/lib/firmware/brcm/brcmfmac4330-sdio.solidrun,cubox-i-q.txt
-rw-r--r--    1 root     root           701 Apr  7  2020 /usr/lib/kernel-overlays/base/lib/firmware/brcm/brcmfmac4330-sdio.txt
-rw-r--r--    1 root     root          1791 Jan 10 21:15 /usr/lib/kernel-overlays/base/lib/firmware/brcm/brcmfmac4330-sdio.wand,imx6dl-wandboard.txt
-rw-r--r--    1 root     root          1791 Jan 10 21:15 /usr/lib/kernel-overlays/base/lib/firmware/brcm/brcmfmac4330-sdio.wand,imx6q-wandboard.txt
```

**Comment by heitbaum** [comment]
> > brcmfmac4330-sdio.solidrun,cubox-i-q.bin
> 
> Harmless, but can be silenced with an alias in the firmware repo.

Something not right… after updating. Still getting the wrong contents in the cubox image.

```
cubox:~ # ls -la /usr/lib/kernel-overlays/base/lib/firmware/brcm/brcmfmac4330-sdio*
-rw-r--r--    1 root     root          1509 Jan 10 21:51 /usr/lib/kernel-overlays/base/lib/firmware/brcm/brcmfmac4330-sdio.Prowise-PT301.txt
-rw-r--r--    1 root     root        222126 Jan 10 21:51 /usr/lib/kernel-overlays/base/lib/firmware/brcm/brcmfmac4330-sdio.bin
-rw-r--r--    1 root     root          1789 Jan 10 21:51 /usr/lib/kernel-overlays/base/lib/firmware/brcm/brcmfmac4330-sdio.solidrun,cubox-i-dl.txt
-rw-r--r--    1 root     root          1789 Jan 10 21:51 /usr/lib/kernel-overlays/base/lib/firmware/brcm/brcmfmac4330-sdio.solidrun,cubox-i-q.txt
-rw-r--r--    1 root     root           701 Apr  7  2020 /usr/lib/kernel-overlays/base/lib/firmware/brcm/brcmfmac4330-sdio.txt
-rw-r--r--    1 root     root          1791 Jan 10 21:51 /usr/lib/kernel-overlays/base/lib/firmware/brcm/brcmfmac4330-sdio.wand,imx6dl-wandboard.txt
-rw-r--r--    1 root     root          1791 Jan 10 21:51 /usr/lib/kernel-overlays/base/lib/firmware/brcm/brcmfmac4330-sdio.wand,imx6q-wandboard.txt

cubox:~ # dmesg | grep brcmfmac
[   16.429946] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac4330-sdio for chip BCM4330/4
[   16.432722] brcmfmac mmc0:0001:1: Direct firmware load for brcm/brcmfmac4330-sdio.solidrun,cubox-i-q.bin failed with error -2
[   16.907927] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac4330-sdio for chip BCM4330/4
[   16.908407] brcmfmac: brcmf_c_process_clm_blob: no clm_blob available (err=-2), device may have limited channels available
[   16.917601] brcmfmac: brcmf_c_preinit_dcmds: Firmware: BCM4330/4 wl0: Oct 25 2011 19:34:12 version 5.90.125.104
```

**Comment by heitbaum** [comment]
> > brcmfmac4330-sdio.solidrun,cubox-i-q.bin
> 
> Harmless, but can be silenced with an alias in the firmware repo.

@chewitt - Fixed with your update + https://github.com/LibreELEC/LibreELEC.tv/pull/6103/commits/a12b947a3976981cf276e99b531e3f0a702ece7d

```
cubox:~ # dmesg | grep brcm
[   16.357142] Bluetooth: hci0: BCM4330B1 'brcm/BCM4330B1.hcd' Patch
[   16.608627] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac4330-sdio for chip BCM4330/4
[   17.206679] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac4330-sdio for chip BCM4330/4
[   17.207823] brcmfmac: brcmf_c_process_clm_blob: no clm_blob available (err=-2), device may have limited channels available
[   17.209434] brcmfmac: brcmf_c_preinit_dcmds: Firmware: BCM4330/4 wl0: Oct 25 2011 19:34:12 version 5.90.125.104
```

**Comment by jernejsk** [comment]
looks good on Allwinner H6, I guess other AW SoCs work too

**Review by CvH — APPROVED** [review]

**Pull request merged**